### PR TITLE
Migrate S4158: Support Constructor with Enumerable

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Part of #7300

### CODE REVIEW

File: analyzers/its/expected/akka.net/Akka.Cluster.Sharding--netstandard2.0-S4158.json [hunk 1]
```diff
@@ -1,17 +0,0 @@
...
-"region":  {
-"startLine":  1305,
-"startColumn":  36,
-"endLine":  1305,
-"endColumn":  46
-}
-}
-}
-]
-}
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Cluster.Sharding--netstandard2.0-S4158.json:17]: FP

File: analyzers/its/expected/akka.net/Akka.Persistence--netstandard2.0-S4158.json [hunk 2]
```diff
@@ -1,17 +0,0 @@
...
-"region":  {
-"startLine":  316,
-"startColumn":  26,
-"endLine":  316,
-"endColumn":  61
-}
-}
-}
-]
-}
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Persistence--netstandard2.0-S4158.json:17]: FP

File: analyzers/its/expected/akka.net/Akka.Streams.TestKit--netstandard2.0-S4158.json [hunk 3]
```diff
@@ -1,43 +0,0 @@
...
-"region":  {
-"startLine":  278,
-"startColumn":  21,
-"endLine":  278,
-"endColumn":  44
-}
-}
-}
-]
-}
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.Streams.TestKit--netstandard2.0-S4158.json:43]: FP

File: analyzers/its/expected/akka.net/Akka.TestKit--netstandard2.0-S4158.json [hunk 4]
```diff
@@ -1,17 +0,0 @@
...
-"region":  {
-"startLine":  234,
-"startColumn":  24,
-"endLine":  234,
-"endColumn":  44
-}
-}
-}
-]
-}
```
Comment (Tim-Pohlmann) [inline analyzers/its/expected/akka.net/Akka.TestKit--netstandard2.0-S4158.json:17]: FP

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.cs [hunk 5]
```diff
@@ -59,17 +59,19 @@ public void ConstructorWithCapacity()
     public void ConstructorWithEnumerable()
     {
         var list = new List<int>(items);
-        list.Clear();   // Noncompliant FP
+        list.Clear();   // Compliant
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.cs:62]: What if...
```
var empty = new List<int>();
var list = new List<int>(empty);
list.Clear();   // Noncompliant
```
We could propagate the constraint, if it exists.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 6]
```diff
@@ -138,6 +138,11 @@ protected override ProgramState PreProcessSimple(SymbolicContext context)
         return state;
     }
 
+    private static bool IsEmptyObject(IOperation operation) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:141]: Because it looks at `TrackedCollectionTypes`
```suggestion
    private static bool IsEmptyCollection(IOperation operation) =>
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 7]
```diff
@@ -138,6 +138,11 @@ protected override ProgramState PreProcessSimple(SymbolicContext context)
...
+        operation.AsObjectCreation() is { } objectCreation
+        && objectCreation.Type.IsAny(TrackedCollectionTypes)
+        && objectCreation.Constructor.Parameters.All(x => !x.Type.DerivesOrImplements(KnownType.System_Collections_IEnumerable));
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:144]: Why `All`? This should look at 1st argument type only, if present. Shouldn't it? Is there a constructor overload that accepts the argument at 2nd place?

Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:144]: This could propagate Empty constraint when present on the 1st argument (of correct type)

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
This will be a bit more trickier

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 8]
```diff
@@ -115,9 +115,9 @@ public abstract class EmptyCollectionsShouldNotBeEnumeratedBase : SymbolicRuleCh
     protected override ProgramState PreProcessSimple(SymbolicContext context)
     {
         var operation = context.Operation.Instance;
-        if (operation.AsObjectCreation()?.Type.IsAny(TrackedCollectionTypes) ?? false)
+        if (operation.AsObjectCreation() is { } objectCreation && objectCreation.Type.IsAny(TrackedCollectionTypes))
         {
-            return context.State.SetOperationConstraint(operation, CollectionConstraint.Empty);
+            return SetConstraintForObjectCreation(context.State, objectCreation);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:120]:
```suggestion
            return SetConstraintForCollectionCreation(context.State, objectCreation);
```
or
```suggestion
            return SetConstraintForCollectionConstructor(context.State, objectCreation);
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 9]
```diff
@@ -147,4 +147,23 @@ public override void ExecutionCompleted()
             ReportIssue(operation, operation.Syntax.ToString());
         }
     }
+
+    private static ProgramState SetConstraintForObjectCreation(ProgramState state, IObjectCreationOperationWrapper objectCreation) =>
+        FirstArgumentAsCollection(objectCreation) switch
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:152]: To give it name based on what it gives us, instead of description of how it does that. 
And it doesn't need to be the 1st one when calling stuff out of order.
```suggestion
        CollectionArgument(objectCreation) switch
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 10]
```diff
@@ -147,4 +147,23 @@ public override void ExecutionCompleted()
...
+        {
+            { } firstArgument when ConstraintFromArgument(state, firstArgument) is { } constraint => state.SetOperationConstraint(objectCreation, constraint),
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:154]: It doens't need to be the 1st one when calling stuff out of order.
```suggestion
            { } collectionArgument when ConstraintFromArgument(state, collectionArgument) is { } constraint => state.SetOperationConstraint(objectCreation, constraint),
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 11]
```diff
@@ -147,4 +147,23 @@ public override void ExecutionCompleted()
...
+    private static IArgumentOperationWrapper? FirstArgumentAsCollection(IObjectCreationOperationWrapper objectCreation) =>
+        objectCreation.Arguments.Select(x => x.AsArgument()).SingleOrDefault(x => x?.Parameter.Ordinal == 0) is { } firstArgument
+        && firstArgument.Parameter.Type.DerivesOrImplements(KnownType.System_Collections_IEnumerable)
+        ? firstArgument
+        : null;
+
+    private static CollectionConstraint ConstraintFromArgument(ProgramState state, IArgumentOperationWrapper argument) =>
+        argument.WrappedOperation.TrackedSymbol() is { } symbol
+        ? state[symbol]?.Constraint<CollectionConstraint>()
+        : null;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:168]: This should not look at symbol, but the operation itself. And after that, this should not be embedded.
```suggestion
        state[argument]?.Constraint<CollectionConstraint>()
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 12]
```diff
@@ -147,4 +147,23 @@ public override void ExecutionCompleted()
...
+            _ => state.SetOperationConstraint(objectCreation, CollectionConstraint.Empty)
+        };
+
+    private static IArgumentOperationWrapper? FirstArgumentAsCollection(IObjectCreationOperationWrapper objectCreation) =>
+        objectCreation.Arguments.Select(x => x.AsArgument()).SingleOrDefault(x => x?.Parameter.Ordinal == 0) is { } firstArgument
+        && firstArgument.Parameter.Type.DerivesOrImplements(KnownType.System_Collections_IEnumerable)
+        ? firstArgument
+        : null;
+
+    private static CollectionConstraint ConstraintFromArgument(ProgramState state, IArgumentOperationWrapper argument) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:165]: Do you have an item somewhere to learn about `Array.Empty` and `Enumerable.Empty` helpers later?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 13]
```diff
@@ -147,4 +147,23 @@ public override void ExecutionCompleted()
             ReportIssue(operation, operation.Syntax.ToString());
         }
     }
+
+    private static ProgramState SetConstraintForObjectCreation(ProgramState state, IObjectCreationOperationWrapper objectCreation) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:151]: The pattern matching makes this very cryptic and hard to follow. Make it `if/else` with one ternary inside instead.

The other comment asks to remove `ConstraintFromArgument` and that will make this way easier to do

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 14]
```diff
@@ -147,4 +147,23 @@ public override void ExecutionCompleted()
...
+        FirstArgumentAsCollection(objectCreation) switch
+        {
+            { } firstArgument when ConstraintFromArgument(state, firstArgument) is { } constraint => state.SetOperationConstraint(objectCreation, constraint),
+            { } => state,
+            _ => state.SetOperationConstraint(objectCreation, CollectionConstraint.Empty)
+        };
+
+    private static IArgumentOperationWrapper? FirstArgumentAsCollection(IObjectCreationOperationWrapper objectCreation) =>
+        objectCreation.Arguments.Select(x => x.AsArgument()).SingleOrDefault(x => x?.Parameter.Ordinal == 0) is { } firstArgument
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:160]: It's not the first one when calling stuff out of order. It's argument of first parameter, but the idea should be that we want the collection argument
```suggestion
        objectCreation.Arguments.Select(x => x.AsArgument()).SingleOrDefault(x => x?.Parameter.Ordinal == 0) is { } collectionArgument
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
Round 2, mostly refactorings and one improvement

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.cs [hunk 15]
```diff
@@ -59,17 +59,38 @@ public void ConstructorWithCapacity()
     public void ConstructorWithEnumerable()
     {
         var list = new List<int>(items);
-        list.Clear();                   // Noncompliant FP
+        list.Clear();                   // Compliant
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyCollectionsShouldNotBeEnumerated.cs:62]: I don't think we have test-cases for it yet. But it's a checkbox in #7300.
I'll add a test for it.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:160]: This requires the line to be re-written like this, which is way harder to read:
`        objectCreation.Arguments.Select(x => x.ToArgument() as IArgumentOperationWrapper?).SingleOrDefault(x => x?.Parameter.Ordinal == 0) is { } firstArgument`
Or do you see an alternative?

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs (hunk 11 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:168]: This does not work. `argument` doesn't seem to have a constraint even in very simple cases.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:160]: Why would you do `as IArgumentOperationWrapper`? The `ToArgument` already returns `IArgumentOperationWrapper`

[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:160]: And it can be further simplified with the idea that we don't need to cast it, returning `IOperation` is enough

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs (hunk 12 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:165]: Now I have. I also add some tests.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs (hunk 11 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:168]: The `return` of `Argument` processor should be likely like this:
```
        return state[argument.Value] is { } value
            ? state.SetOperationValue(argument, value)
            : state;
```

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs (hunk 13 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:151]: I took a different approach. Let me know what you think of it.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 16]
```diff
@@ -147,4 +147,23 @@ public override void ExecutionCompleted()
...
+
+    private static ProgramState SetConstraintForObjectCreation(ProgramState state, IObjectCreationOperationWrapper objectCreation) =>
+        FirstArgumentAsCollection(objectCreation) switch
+        {
+            { } firstArgument when ConstraintFromArgument(state, firstArgument) is { } constraint => state.SetOperationConstraint(objectCreation, constraint),
+            { } => state,
+            _ => state.SetOperationConstraint(objectCreation, CollectionConstraint.Empty)
+        };
+
+    private static IArgumentOperationWrapper? FirstArgumentAsCollection(IObjectCreationOperationWrapper objectCreation) =>
```
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:159]: See the other comment.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs]: It does not work without the `as IArgumentOperationWrapper?` because `SingleOrDefault` will not return `null` for a struct, and there is no easy way to check for `default`.
Omitting the check on `Ordinal` simplifies it enough to work around the problem. I have checked all covered collections, and no constructor takes a different kind of collection as a parameter. So it should be safe.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs (hunk 11 above)
Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs]: This worked, nice. It does not fix the `new List<int>(condition ? empty : empty)` FN, though.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs [hunk 17]
```diff
@@ -115,8 +115,13 @@ public abstract class EmptyCollectionsShouldNotBeEnumeratedBase : SymbolicRuleCh
     protected override ProgramState PreProcessSimple(SymbolicContext context)
     {
         var operation = context.Operation.Instance;
-        if ((operation.AsObjectCreation()?.Type.IsAny(TrackedCollectionTypes) ?? false)
-            || IsEmptyArray(operation))
+        if (operation.AsObjectCreation() is { } objectCreation && objectCreation.Type.IsAny(TrackedCollectionTypes))
+        {
+            return ConstraintFromCollectionCreation(context.State, objectCreation) is { } constraint
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/EmptyCollectionsShouldNotBeEnumeratedBase.cs:120]: I'm not very happy with the name, as it's confusing `objectCreation` vs. the argument value. What about something like `CollectionArgumentConstraint`?

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
One more UT and will be good to go

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Argument.cs [hunk 18]
```diff
@@ -44,6 +44,8 @@ private static ProgramState ProcessArgument(ProgramState state, IArgumentOperati
         {
             state = state.SetSymbolConstraint(notNullSymbol, ObjectConstraint.NotNull);
         }
-        return state;
+        return state[argument.Value] is { } value
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Argument.cs:47]: I'll do this in #7427

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM
